### PR TITLE
support rename table

### PR DESCRIPTION
**Pull request by lsy3993**

### CODE REVIEW

**Comment by lsy3993** [comment]
kitex version is lower than dev branch, just close

File: pkg/ccr/base/spec.go [hunk 1]
```diff
@@ -857,6 +857,36 @@ func (s *Spec) TruncateTable(destTableName string, truncateTable *record.Truncat
 	return s.DbExec(sql)
 }
 
+func (s *Spec) RenameTable(destTableName string, renameTable *record.RenameTable) error {
+	// rename table may be 'rename table', 'rename rollup', 'rename partition' and 'rename column'
```
Comment (smallx) [inline pkg/ccr/base/spec.go:861]: remove `and 'rename column'` due to https://github.com/selectdb/ccr-syncer/pull/139

File: pkg/ccr/base/spec.go [hunk 2]
```diff
@@ -857,6 +857,36 @@ func (s *Spec) TruncateTable(destTableName string, truncateTable *record.Truncat
...
+		sql = fmt.Sprintf("ALTER TABLE %s RENAME ROLLUP %s %s", destTableName, renameTable.OldRollupName, renameTable.NewRollupName)
+	}
+
+	// ALTER TABLE example_table RENAME PARTITION p1 p2;
+	// if rename partition, table name is unchanged
+	if renameTable.NewParitionName != "" && renameTable.OldParitionName != "" {
+		sql = fmt.Sprintf("ALTER TABLE %s RENAME PARTITION %s %s;", destTableName, renameTable.OldParitionName, renameTable.NewParitionName)
+	}
+
+	// todo rename column
```
Comment (smallx) [inline pkg/ccr/base/spec.go:880]: remove this due to https://github.com/selectdb/ccr-syncer/pull/139

**Review by smallx — CHANGES_REQUESTED** [review]

File: pkg/ccr/job.go [hunk 3]
```diff
@@ -1136,6 +1136,37 @@ func (j *Job) handleReplacePartitions(binlog *festruct.TBinlog) error {
...
+	data := binlog.GetData()
+	renameTable, err := record.NewRenameTableFromJson(data)
+	if err != nil {
+		return err
+	}
+
+	// don't support rename table when table sync
+	var destTableName string
+	if j.SyncType == TableSync {
+		return xerror.Errorf(xerror.Normal, "rename table is not supported when table sync")
```
Comment (w41ter) [inline pkg/ccr/job.go:1152]: We don't support sync rename operation to the downstream, ignore this binlog directly, and remember to update the saved src name.

File: pkg/ccr/job.go [hunk 4]
```diff
@@ -1136,6 +1136,37 @@ func (j *Job) handleReplacePartitions(binlog *festruct.TBinlog) error {
...
+		}
+
+		if destTableName, err = j.destMeta.GetTableNameById(destTableId); err != nil {
+			return err
+		} else if destTableName == "" {
+			return xerror.Errorf(xerror.Normal, "tableId %d not found in destMeta", destTableId)
+		}
+	}
+
+	err = j.IDest.RenameTable(destTableName, renameTable)
```
Comment (w41ter) [inline pkg/ccr/job.go:1166]: Once the rename operation is executed success, the cached metadata in destMeta is staled, try refresh it:

```
j.destMeta.GetTables()
```

Comment (w41ter) [reply pkg/ccr/job.go:1166]: the srcMeta should be updated in both db and table sync.

```
j.srcMeta.GetTables()
```

**Comment by lsy3993** [comment]
new PR is : #147

**Comment by lsy3993** [comment]
New PR is https://github.com/selectdb/ccr-syncer/pull/147, close this PR

**Pull request closed without merging**